### PR TITLE
types: Fix `TIMESTAMPADD` lost fractional digits

### LINKED ISSUE #31795: `TIMESTAMPADD`  does not return correct number of fractional digit in datetime expression
### 1. Minimal reproduce step 
```sql
SET sql_mode='';
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (s1 DATETIME(2));
INSERT INTO t1 VALUES (TIMESTAMP '0001-01-01 00:00:00.05');
SELECT TIMESTAMPADD(second, 1, s1) FROM t1;
DROP TABLE t1;
SELECT TIMESTAMPADD(second, 1.1, '2011-05-18 16:17:11.0');
SET sql_mode=default;
```

### 2. What did you expect to see? 
```sql
mysql> SELECT TIMESTAMPADD(second, 1, s1) FROM t1;
+-----------------------------+
| TIMESTAMPADD(second, 1, s1) |
+-----------------------------+
| 0001-01-01 00:00:01.05      |
+-----------------------------+

mysql> SELECT TIMESTAMPADD(second, 1.1, '2011-05-18 16:17:11.0');
+----------------------------------------------------+
| TIMESTAMPADD(second, 1.1, '2011-05-18 16:17:11.0') |
+----------------------------------------------------+
| 2011-05-18 16:17:12.100000                         |
+----------------------------------------------------+
```

### 3. What did you see instead 
```sql
tidb> SELECT TIMESTAMPADD(second, 1, s1) FROM t1;
+-----------------------------+
| TIMESTAMPADD(second, 1, s1) |
+-----------------------------+
| 0001-01-01 00:00:01         |
+-----------------------------+

tidb> SELECT TIMESTAMPADD(second, 1.1, '2011-05-18 16:17:11.0');
+----------------------------------------------------+
| TIMESTAMPADD(second, 1.1, '2011-05-18 16:17:11.0') |
+----------------------------------------------------+
| 2011-05-18 16:17:12                                |
+----------------------------------------------------+
```
### 4. What is your TiDB version? 
```
tidb_version(): Release Version: v5.5.0-alpha-105-gaabd4e04d
Edition: Community
Git Commit Hash: aabd4e04d994eb91663abaa80865daec4cf970a6
Git Branch: master
UTC Build Time: 2022-01-13 05:36:36
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by birdstorm**
Signed-off-by: birdstorm <samuelwyf@hotmail.com>

### What problem does this PR solve?

Issue Number: close #31795 

Problem Summary:

### What is changed and how it works?

TiDB lost fractional digits when using `SECOND` as _**unit** type_ for _interval_. [Refrence](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_timestampadd)

In the current `mysql-test`, `TIMESTAMPADD()` test output does not match the result of MySQL(version 8.0, 5.7, 5.6, 5.5)

- Since 8.0: overflow on DateTime types in `TIMESTAMPADD()` return null regardless of _**unit**_, with a warning indicating DateTime type overflow. When using _**unit**_ `second`, the fractional part of seconds is preserved.
- Since 5.6: overflow on DateTime types in `TIMESTAMPADD()` returns null and output warnings when using _**unit**_ `quarter`, `month`, `week`, `day`, `microsecond`. When using _**unit**_ `second`, the fractional part of seconds is preserved.
- Since 5.5: overflow on DateTime types in `TIMESTAMPADD()` returns null and output warnings when using _**unit**_ `quarter`, `month`, `week`, `day`, `second`, `microsecond`. When using _**unit**_ `second`, the fractional part of seconds is **omitted**.

In this PR, when `TIMESTAMPADD()` overflows, it will follow MySQL 8.0's behavior and return null for all _**unit**_. When using _**unit**_ `second`, the fractional part of seconds will be **PRESERVED**.

Although the MySQL document states that `TIMESTAMPADD(unit, interval, datetime_expr)` ensures `interval` is an integer expression, the result of test cases on MySQL(version 8.0, 5.7, 5.6, 5.5) shows exceptions.

e.g., 
```sql
mysql> select timestampadd(month,1.5,'2000-01-01 00:00:00') as result;
+---------------------+
| result              |
+---------------------+
| 2000-03-01 00:00:00 |
+---------------------+
1 row in set (0.00 sec)

mysql> select timestampadd(second,1.5,'2000-01-01 00:00:00') as result;
+----------------------------+
| result                     |
+----------------------------+
| 2000-01-01 00:00:01.500000 |
+----------------------------+
1 row in set (0.00 sec)

mysql> select timestampadd(microsecond,1.5,'2000-01-01 00:00:00') as result;
+----------------------------+
| result                     |
+----------------------------+
| 2000-01-01 00:00:00.000002 |
+----------------------------+
1 row in set (0.00 sec)
```

- When `unit` is `second`, the interval can be of float types. I guess it is because it is transformed into equivalent representation of `microsecond`.
- In other cases, the interval will be rounded to nearest integer.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

File: expression/builtin_time.go [hunk 1]
```diff
@@ -6120,14 +6120,90 @@ func (b *builtinTimestampAddSig) Clone() builtinFunc {
 	return newSig
 }
 
+var (
+	minDatetimeInGoTime, _ = types.MinDatetime.GoTime(time.Local)
+	minDatetimeNanos       = float64(minDatetimeInGoTime.Unix())*1e9 + float64(minDatetimeInGoTime.Nanosecond())
+	maxDatetimeInGoTime, _ = types.MaxDatetime.GoTime(time.Local)
+	maxDatetimeNanos       = float64(maxDatetimeInGoTime.Unix())*1e9 + float64(maxDatetimeInGoTime.Nanosecond())
+	minDatetimeMonths      = float64(1000*12 + 0)
```
Comment (windtalker) [inline expression/builtin_time.go:6128]: Why `minDatetimeMonths` starts from 12000?

Comment (birdstorm) [reply expression/builtin_time.go:6128]: oh, that is not correct, it should be the months of 0001-01-01, which is 12 months since year 0000.

Comment (birdstorm) [reply expression/builtin_time.go:6128]: done

**Review by windtalker — APPROVED** [review]
LGTM

**Comment by birdstorm** [comment]
In response to MergeCI:
- `common-test` and `integration-common-test` relates to `mysql-test pr#2018`, and should be fixed now after merging.
- other test failures are not related to this PR

**Pull request merged**